functions/go/set-namespace/README.md

```diff
@@ -17,68 +17,114 @@ Namespaces are often used in the following scenarios:
-subjects:
-  - kind: ServiceAccount
-    name: default # <================== name default is used
-    namespace: example # <================== this will be updated
-roleRef:
-  kind: Role
-  name: confluent-operator
-  apiGroup: rbac.authorization.k8s.io
+kind: ConfigMap
+data:
+  namespace: newNamespace # required
+  namespaceSelector: example # update namespace whose value is "example" to "newNamespace"
 ```
 
-This function can be used both declaratively and imperatively.
+SetNamespace as functionConfig
+```yaml
+apiVersion: fn.kpt.dev/v1alpha1
+kind: SetNamespace
+namespace: newNamespace # required
+namespaceSelector: example # update namespace whose value is "example" to "newNamespace"
+```
 
-### FunctionConfig
 
-There are 2 kinds of `functionConfig` supported by this function:
+### Three updating modes
 
-- `ConfigMap`
-- A custom resource of kind `SetNamespace`
+This function supports three modes to flexibly choose and update the target namespaces.
 
-To use a `ConfigMap` as the `functionConfig`, the desired namespace must be
-specified in the `data.namespace` field.
+##### Restrict Mode
+All target KRM resources namespace has to have the same value. All namespace will be updated to the new value. 
 
-To add a namespace `staging` to all resources, we use the
-following `functionConfig`:
+ConfigMap as functionConfig
+```yaml
+apiVersion: v1
+kind: ConfigMap
+data:
+  namespace: newNamespace # update all namespace fields to "newNamespace"
+```
+
+##### DefaultNamespace Mode
+
+The input `resourcelist.items` contains one and only one namespace object. The function matches the namespace `metadata.name`
+with all other KRM resources, and only update the namespace if it matches the namespace object. 
+If more than one namespace objects are found, raise errors;
 
+```yaml
+kind: ResourceList
+functionConfig:
+  apiVersion: v1
+  kind: ConfigMap
+  data:
+    namespace: newNs
+items:
+- apiVersion: v1
+  kind: Namespace
+  metadata:
+    name: example # updated to "newNs"
+- apiVersion: v1
+  kind: Service
+  metadata:
+    name: the-service1
+    namespace: example  # updated to "newNs"
+- apiVersion: v1
+  kind: Service
+  metadata:
+    name: the-service2
+    namespace: irrelevant # skip since namespace does not match "example".
+```
+
+##### Selector Mode
+
+Only updates the namespace which matches a given value. The "oldNamespace" refers to the argument in FunctionConfig
+
+ConfigMap as functionConfig
 ```yaml
 apiVersion: v1
 kind: ConfigMap
-metadata:
-  name: my-config
 data:
-  namespace: staging
+  namespace: newNamespace
+  namespaceSelector: example # update namespace whose value is "example" to "newNamespace"
+```
+
+SetNamespace as functionConfig
+```yaml
+apiVersion: fn.kpt.dev/v1alpha1
+kind: SetNamespace
+namespace: newNamespace
+namespaceSelector: example # update namespace whose value is "example" to "newNamespace"
```

Comment:
Maybe `namespaceMatcher` or `originalNamespace` is a better name.
You may want to hear other folks' opinion.

Reply:
Sure, I prefer `namespaceMatcher` over `orignialNamespace`. The latter doesn't tell why the functions need the orignal namespace.

Reply:
looks like [`namespaceSelector`](https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors) indeed can confuse users.